fabric-model-loading-api-v1/src/client/java/net/fabricmc/fabric/impl/client/model/loading/ModelLoadingEventDispatcher.java

```diff
@@ -0,0 +1,465 @@
+
+		modelResolverContextStack.push(context);
+		return model;
+	}
+
+	public UnbakedModel modifyModelOnLoad(Identifier id, UnbakedModel model) {
+		if (onLoadModifierContextStack.isEmpty()) {
+			onLoadModifierContextStack.add(new OnLoadModifierContext());
+		}
+
+		OnLoadModifierContext context = onLoadModifierContextStack.pop();
+		context.prepare(id);
+
+		model = pluginContext.modifyModelOnLoad().invoker().modifyModelOnLoad(model, context);
+
+		onLoadModifierContextStack.push(context);
+		return model;
+	}
+
+	public UnbakedModel modifyModelBeforeBake(UnbakedModel model, Identifier id, Function<SpriteIdentifier, Sprite> textureGetter, ModelBakeSettings settings, Baker baker) {
+		if (beforeBakeModifierContextStack.isEmpty()) {
+			beforeBakeModifierContextStack.add(new BeforeBakeModifierContext());
+		}
+
+		BeforeBakeModifierContext context = beforeBakeModifierContextStack.pop();
+		context.prepare(id, textureGetter, settings, baker);
+
+		model = pluginContext.modifyModelBeforeBake().invoker().modifyModelBeforeBake(model, context);
+
+		beforeBakeModifierContextStack.push(context);
+		return model;
+	}
+
+	public BakedModel modifyModelAfterBake(BakedModel model, Identifier id, UnbakedModel sourceModel, Function<SpriteIdentifier, Sprite> textureGetter, ModelBakeSettings settings, Baker baker) {
+		if (afterBakeModifierContextStack.isEmpty()) {
+			afterBakeModifierContextStack.add(new AfterBakeModifierContext());
+		}
+
+		AfterBakeModifierContext context = afterBakeModifierContextStack.pop();
+		context.prepare(id, sourceModel, textureGetter, settings, baker);
+
+		model = pluginContext.modifyModelAfterBake().invoker().modifyModelAfterBake(model, context);
+
+		afterBakeModifierContextStack.push(context);
+		return model;
+	}
+
+	private class ModelResolverContext implements ModelResolver.Context {
+		private Identifier id;
+
+		private void prepare(Identifier id) {
+			this.id = id;
+		}
+
+		@Override
+		public Identifier id() {
+			return id;
+		}
+
+		@Override
+		public UnbakedModel getOrLoadModel(Identifier id) {
+			return ((ModelLoaderHooks) loader).fabric_getOrLoadModel(id);
+		}
+
+		@Override
+		public ModelLoader loader() {
+			return loader;
+		}
+	}
+
+	private class BlockStateResolverContext implements BlockStateResolver.Context {
+		private Block block;
+		private final Reference2ReferenceMap<BlockState, UnbakedModel> models = new Reference2ReferenceOpenHashMap<>();
+
+		private void prepare(Block block) {
+			this.block = block;
+			models.clear();
+		}
+
+		@Override
+		public Block block() {
+			return block;
+		}
+
+		@Override
+		public void setModel(BlockState state, UnbakedModel model) {
+			Objects.requireNonNull(model, "state cannot be null");
+			Objects.requireNonNull(model, "model cannot be null");
+
+			if (!state.isOf(block)) {
+				throw new IllegalArgumentException("Attempted to set model for state " + state + " on block " + block);
+			}
+
+			if (models.put(state, model) != null) {
```

Comment:
```suggestion
			if (models.putIfAbsent(state, model) != null) {
```
We use `putIfAbsent` for similar code in other modules